libyul/optimiser/BlockHasher.cpp

```diff
@@ -0,0 +1,193 @@
+	return result;
+}
+
+void BlockHasher::operator()(Literal const& _literal)
+{
+	hash64(compileTimeLiteralHash("Literal"));
+	hash64(_literal.value.hash());
+	hash64(_literal.type.hash());
+	hash8(static_cast<uint8_t>(static_cast<std::underlying_type_t<LiteralKind>>(_literal.kind)));
+}
+
+void BlockHasher::operator()(Identifier const& _identifier)
+{
+	hash64(compileTimeLiteralHash("Identifier"));
+	auto it = m_variableReferences.find(_identifier.name);
+	if (it == m_variableReferences.end())
+	{
+		it = m_variableReferences.emplace(_identifier.name, VariableReference {
+			m_externalIdentifierCount++,
+			true
+		}).first;
+		m_externalReferences.emplace_back(_identifier.name);
+	}
+	if (it->second.isExternal)
+		hash64(compileTimeLiteralHash("external"));
+	else
+		hash64(compileTimeLiteralHash("internal"));
+	hash64(it->second.id);
+}
+
+void BlockHasher::operator()(FunctionalInstruction const& _instr)
+{
+	hash64(compileTimeLiteralHash("FunctionalInstruction"));
+	hash8(static_cast<std::underlying_type_t<eth::Instruction>>(_instr.instruction));
+	hash64(_instr.arguments.size());
+	ASTWalker::operator()(_instr);
+}
+
+void BlockHasher::operator()(FunctionCall const& _funCall)
+{
+	hash64(compileTimeLiteralHash("FunctionCall"));
+	hash64(_funCall.functionName.name.hash());
+	hash64(_funCall.arguments.size());
+	ASTWalker::operator()(_funCall);
+}
+
+void BlockHasher::operator()(ExpressionStatement const& _statement)
+{
+	hash64(compileTimeLiteralHash("ExpressionStatement"));
+	ASTWalker::operator()(_statement);
+}
+
+void BlockHasher::operator()(Assignment const& _assignment)
+{
+	hash64(compileTimeLiteralHash("Assignment"));
+	hash64(_assignment.variableNames.size());
+	for (auto const& name: _assignment.variableNames)
+		(*this)(name);
+	visit(*_assignment.value);
+}
+
+void BlockHasher::operator()(VariableDeclaration const& _varDecl)
+{
+	hash64(compileTimeLiteralHash("VariableDeclaration"));
+	hash64(_varDecl.variables.size());
+	for (auto const& var: _varDecl.variables)
+	{
+		yulAssert(!m_variableReferences.count(var.name), "");
+		m_variableReferences[var.name] = VariableReference{
+			m_internalIdentifierCount++,
+			false
+		};
+	}
+	ASTWalker::operator()(_varDecl);
+}
+
+void BlockHasher::operator()(If const& _if)
+{
+	hash64(compileTimeLiteralHash("If"));
+	ASTWalker::operator()(_if);
+}
+
+void BlockHasher::operator()(Switch const& _switch)
+{
+	hash64(compileTimeLiteralHash("Switch"));
+	hash64(_switch.cases.size());
+	// Instead of sorting we could consider to combine
+	// the case hashes using a commutative operation here.
+	std::set<Case const*, SwitchCaseSorter> cases;
+	for (auto const& _case: _switch.cases)
+		cases.insert(&_case);
+
+	visit(*_switch.expression);
+	for (auto const& _case: cases)
+	{
+		if (_case->value)
+			(*this)(*_case->value);
+		(*this)(_case->body);
+	}
+}
+
+void BlockHasher::operator()(FunctionDefinition const& _funDef)
+{
+	hash64(compileTimeLiteralHash("FunctionDefinition"));
```

Comment:
Don't you need to consider function parameters specially?

Reply:
We could hash them specially here, yes - the way it's done now, they'll occur as "external references" in the body block hash later, if used, and won't occur at all, if unused - doesn't really matter, though, since it's just used as heuristics. I don't see a reason for not keeping it as it is for now given the use case, though.